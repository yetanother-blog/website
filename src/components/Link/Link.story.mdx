import { Story, Meta, Preview, Props } from '@storybook/addon-docs/blocks';
import { Link } from './Link';
import { ThemeContext } from 'styled-components';
import { Link as GatsbyLink } from 'gatsby';

<Meta title="Components/Link" />

# Link

Link component to wrap pages and posts.

<Preview>
  <Story name="default">
    <ThemeContext.Consumer>
      {(theme) => (
        <React.Fragment>
          <Link
            variant="primary"
            display="block"
            marginBottom={theme.space.xl}
            href="https://twitter.com/_yetanotherblog"
            target="_blank"
          >
            hello link
          </Link>
          <Link variant="secondary" display="block" marginBottom={theme.space.xl} component={GatsbyLink} to="/lol">
            hello link
          </Link>
          <Link variant="tertiary" display="block" marginBottom={theme.space.xl} component={GatsbyLink} to="/lol">
            hello link
          </Link>
          <Link variant="quaternary" display="block" marginBottom={theme.space.xl} component={GatsbyLink} to="/lol">
            hello link
          </Link>
        </React.Fragment>
      )}
    </ThemeContext.Consumer>
  </Story>
</Preview>

## Props

<Props of={Link} />
