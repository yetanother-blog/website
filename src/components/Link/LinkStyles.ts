import styled, { css } from 'styled-components';

export const sharedStyles = css`
  transition: color 200ms ease-in;
  text-decoration: none;
  outline: none;
  -webkit-font-smoothing: antialiased;
  font-weight: 600;
  cursor: pointer;
  font-family: 'Source Code Pro';

  &:hover,
  &:focus {
    span {
      &:after {
        transform: scaleY(4);
      }
    }
  }

  span {
    &:after {
      content: '';
      position: absolute;
      width: 100%;
      background-color: ${(props) => props.theme.colors.grey300};
      height: 4px;
      bottom: -8px;
      left: 0;
      z-index: -1;
      transform-origin: bottom;
      transition: transform 100ms ease-in;
    }
  }
`;

export const defaultFontLinkStyles = css`
  color: ${(props) => props.theme.colors.grey500};
  font-size: ${(props) => props.theme.fontSizes[3]};
  font-weight: 600;
  letter-spacing: 0.5px;
  line-height: 1.25;
  margin-bottom: 0;
`;

export const smallFontLinkStyles = css`
  font-size: ${(props) => props.theme.fontSizes[2]};
  font-weight: 600;
  line-height: 1.25;
  margin-bottom: 0;
`;

export const tinyFontLinkStyles = css`
  font-size: ${(props) => props.theme.fontSizes[1]};
  font-weight: 600;
  line-height: 1.25;
  margin-bottom: 0;
`;

export const StyledUnderline = styled.span`
  position: relative;

  &:after {
    content: '';
    position: absolute;
    width: 100%;
    background-color: ${(props) => props.theme.colors.grey300};
    height: 4px;
    bottom: -8px;
    left: 0;
    z-index: -1;
    transform-origin: bottom;
    transition: transform 100ms ease-in-out;
  }

  &:hover,
  &:focus {
    &:after {
      transform: scaleY(4);
    }
  }
`;
