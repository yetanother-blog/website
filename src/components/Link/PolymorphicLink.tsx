import React from 'react';

export type PolymorphicLinkProps<C extends React.ElementType> = {
  component: C;
} & React.ComponentPropsWithRef<C>;
export function PolymorphicLink<C extends React.ElementType = 'a'>(props: PolymorphicLinkProps<C>) {
  const { display, marginBottom, component: Component, ...rest } = props;
  return <Component {...rest} />;
}
PolymorphicLink.defaultProps = {
  component: 'a',
};
