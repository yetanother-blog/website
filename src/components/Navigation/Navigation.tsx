import React, { useState, useContext } from 'react';
import styled, { useTheme } from 'styled-components';
import {
  size,
  WidthProps,
  HeightProps,
  layout,
  LayoutProps,
  flexbox,
  FlexboxProps,
  height,
  display,
  padding,
  margin,
  PaddingProps,
  MarginBottomProps,
} from 'styled-system';
import { Logo } from '../Logo/Logo';
import { MobileNavigationContext } from '../../context/MobileNavigationContext/MobileNavigationContext';

interface HamburgerProps {
  isActive: boolean;
}

const Hamburger = styled.div<HamburgerProps>`
  position: relative;
  width: 24px;
  ${flexbox}

  &:before,
  &:after {
    content: '';
    position: absolute;
    left: 0;
    height: 3px;
    width: 24px;
    background-color: ${(props) => props.theme.colors.grey500};
    transform-origin: center center;
    transition: transform ease-in-out 120ms;
  }

  &:after {
    top: ${(props) => (props.isActive ? 0 : '4px')};
    transform: rotate(${(props) => (props.isActive ? '-45deg' : '0deg')});
  }
  &:before {
    top: ${(props) => (props.isActive ? 0 : '-4px')};
    transform: rotate(${(props) => (props.isActive ? '45deg' : '0deg')});
  }
`;

const StyledMobileNavigationButton = styled.button<LayoutProps & FlexboxProps>`
  ${layout}
  ${flexbox}
  padding: 0;
  background-color: transparent;
  outline: none;
  border: none;
  z-index: 100;
  position: absolute;
  top: 10px;
  right: 15px;
`;

const StyledNav = styled.nav<
  WidthProps & HeightProps & LayoutProps & FlexboxProps & PaddingProps & MarginBottomProps
>`
  ${layout}
  ${height}
  ${display}
  ${flexbox}
  ${size}
  ${padding}
  ${margin}
`;

const StyledMobileWrapper = styled.div`
  position: fixed;
  width: 100%;
  height: 100vh;
  top: 0;
  left: 0;
  background-color: white;
  z-index: 10;
  display: flex;
  flex-direction: column;
  padding: 60px 30px;
  align-items: flex-end;

  > * {
    margin-bottom: 20px;
  }
`;

const StyledDesktopWrapper = styled.div<LayoutProps>`
  ${layout}
  align-items: center;

  > * {
    margin-right: ${(props) => props.theme.space.l}px;
    &:last-child {
      margin-right: 0;
    }
  }
`;

export const Navigation: React.FC = ({ children }) => {
  const { isMobileNavigationOpen, setIsMobileNavigationOpen } = useContext(MobileNavigationContext);
  const theme = useTheme();

  return (
    <StyledNav
      display="flex"
      pt={theme.space.l}
      pb={theme.space.l}
      alignItems="center"
      justifyContent="space-between"
      height={[64, 80]}
      mb={[theme.space.xxl, theme.space.xxxl]}
      width="100%"
    >
      <Logo to="/" />
      {isMobileNavigationOpen && <StyledMobileWrapper>{children}</StyledMobileWrapper>}
      <StyledDesktopWrapper display={['none', 'flex']}>{children}</StyledDesktopWrapper>
      <StyledMobileNavigationButton
        display={['flex', 'none']}
        height="48px"
        width="48px"
        justifyContent="center"
        alignItems="center"
        onClick={() => setIsMobileNavigationOpen(!isMobileNavigationOpen)}
      >
        <Hamburger isActive={isMobileNavigationOpen} />
      </StyledMobileNavigationButton>
    </StyledNav>
  );
};
