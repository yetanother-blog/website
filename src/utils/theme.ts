import { DefaultTheme } from 'styled-components';

export const theme: DefaultTheme = {
  fontSizes: ['12px', '14px', '16px', '20px', '24px', '32px'],
  colors: {
    primary: '#00FF00',
    secondary: '#FB00FF',
    grey500: '#1C1C1C',
    grey400: '#ACADAC',
    grey300: '#F6F6F6',
  },
  breakpoints: ['576px', '768px', '992px', '1200px'],
  space: {
    zero: 0,
    xxs: 4,
    xs: 8,
    s: 12,
    m: 16,
    l: 20,
    xl: 40,
    xxl: 60,
    xxxl: 120,
  },
  radii: {
    s: 3,
  },
  width: {
    m: '620px',
    l: '940px',
    xl: '1200px',
  },
};
