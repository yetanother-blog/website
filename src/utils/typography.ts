import { CSSObject } from "styled-components";
import Typography from "typography";

const typography = new Typography({
  googleFonts: [
    {
      name: "Source Sans Pro",
      styles: ["400", "600"]
    },
    {
      name: "Source Code Pro",
      styles: ["600", "700"]
    }
  ],
  baseFontSize: "20px",
  baseLineHeight: 1.85,
  bodyFontFamily: ["Source Sans Pro", "sans-serif"],
  boldWeight: "600",
  overrideStyles: ({ adjustFontSizeTo }, options, styles) => ({
    a: {
      fontFamily: `"Source Code Pro", "sans-serif"`
    }
  })
});

// Hot reload typography in development.
if (process.env.NODE_ENV !== `production`) {
  typography.injectStyles();
}

export default typography;

export const rhythm = typography.rhythm;
export const scale = typography.scale;

type StyledScale = (values: number) => CSSObject;
export const styledScale = scale as StyledScale;
