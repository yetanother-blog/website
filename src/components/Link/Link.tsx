import React from 'react';
import styled from 'styled-components';
import { PolymorphicLink, PolymorphicLinkProps } from './PolymorphicLink';
import {
  sharedStyles,
  defaultFontLinkStyles,
  smallFontLinkStyles,
  StyledUnderline,
  tinyFontLinkStyles,
} from './LinkStyles';
import { display, DisplayProps, margin, MarginProps } from 'styled-system';

export type LinkVariant = 'primary' | 'secondary' | 'tertiary' | 'quaternary';

export type LinkProps<C extends React.ElementType> = PolymorphicLinkProps<C> &
  DisplayProps &
  MarginProps & {
    variant?: LinkVariant;
    children: React.ReactNode;
  };

const StyledPrimaryLink = styled(PolymorphicLink)`
  ${display}  
  ${sharedStyles}
  ${defaultFontLinkStyles}
  ${margin}

  &:before {
    content: '>> @';
    color: ${(props) => props.theme.colors.primary};
  }
`;

const StyledSecondaryLink = styled(PolymorphicLink)`
  ${display}
  ${sharedStyles}
  ${defaultFontLinkStyles}
  ${margin}

  &:before {
    content: '>> @';
    color: ${(props) => props.theme.colors.secondary};
  }
`;

const StyledTertiaryLink = styled(PolymorphicLink)`
  ${display}
  ${sharedStyles}
  ${smallFontLinkStyles}
  ${margin}
  color: ${(props) => props.theme.colors.grey500};

  @keyframes pulse {
    0% {
      transform: scaleY(4);
    }
    100% {
      transform: scale(1.3, 8);
    }

  }

  &[aria-current="page"] {
    span {
      &:after {
        animation: pulse 1000ms forwards;
        animation-delay: 300ms;
        transform: scaleY(4);

      }
    }
  }
`;

const StyledQuaternaryLink = styled(PolymorphicLink)`
  ${display}
  ${sharedStyles}
  ${tinyFontLinkStyles}
  ${margin}
  color: ${(props) => props.theme.colors.grey400};
`;

export function Link<C extends React.ElementType>(props: LinkProps<C>) {
  const { variant = 'primary', children, ...restProps } = props;

  switch (variant) {
    case 'primary':
      return (
        <StyledPrimaryLink {...restProps}>
          <StyledUnderline>{children}</StyledUnderline>
        </StyledPrimaryLink>
      );

    case 'secondary':
      return (
        <StyledSecondaryLink {...restProps}>
          <StyledUnderline>{children}</StyledUnderline>
        </StyledSecondaryLink>
      );

    case 'tertiary':
      return (
        <StyledTertiaryLink {...restProps}>
          <StyledUnderline>{children}</StyledUnderline>
        </StyledTertiaryLink>
      );

    default:
      return (
        <StyledQuaternaryLink {...restProps}>
          <StyledUnderline>{children}</StyledUnderline>
        </StyledQuaternaryLink>
      );
  }
}
