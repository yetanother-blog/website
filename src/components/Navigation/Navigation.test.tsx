import React from 'react';
import { render } from '@testing-library/react';
import { ThemeContext } from 'styled-components';
import { theme } from '../../utils/theme';
import { Navigation } from './Navigation';
import { Link } from '../Link/Link';

describe('Navigation component', () => {
  it('should render social link for twitter', () => {
    const { container } = render(
      <ThemeContext.Provider value={theme}>
        <Navigation>
          <Link variant="tertiary">Guide</Link>
        </Navigation>
      </ThemeContext.Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
