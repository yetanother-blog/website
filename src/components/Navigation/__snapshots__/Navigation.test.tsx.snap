// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component should render social link for twitter 1`] = `
<div>
  <nav
    class="sc-jrQzUz NhGfT"
    display="flex"
    height="64,80"
    width="100%"
  >
    <a
      aria-label="logo"
      class="sc-gKckTs cbXYIC"
      href="/"
      role="navigation"
    >
      <p
        class="sc-hKwCoD kVRiJn sc-iCfLBT buuvTd"
        font-family="Source Code Pro"
        font-weight="600"
      >
        yetanother.blog
      </p>
    </a>
    <div
      class="sc-iqsfdx csHdKK"
      display="none,flex"
    >
      <a
        class="sc-ksdxAp eizrhv"
      >
        <span
          class="sc-crHlIS gtoVmF"
        >
          Guide
        </span>
      </a>
    </div>
    <button
      aria-label="hamburger menu"
      class="sc-pVTma gSKnEc"
      display="flex,none"
      height="48px"
      width="48px"
    >
      <div
        class="sc-furvIG iMnDab"
      />
    </button>
  </nav>
</div>
`;
